go/internal/doc-snippets/rag/main.go

```diff
@@ -0,0 +1,227 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rag
+
+import (
+	"bytes"
+	"context"
+	"log"
+
+	"github.com/firebase/genkit/go/ai"
+	"github.com/firebase/genkit/go/genkit"
+	"github.com/firebase/genkit/go/plugins/localvec"
+	"github.com/firebase/genkit/go/plugins/vertexai"
+
+	"github.com/ledongthuc/pdf"
+	"github.com/tmc/langchaingo/textsplitter"
+)
+
+func main() {
+	//!+vec
+	ctx := context.Background()
+
+	err := vertexai.Init(ctx, "", "")
+	if err != nil {
+		log.Fatal(err)
+	}
+	err = localvec.Init()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	menuPdfIndexer, _, err := localvec.DefineIndexerAndRetriever(
+		"menuQA",
+		localvec.Config{
+			Embedder: vertexai.Embedder("text-embedding-004"),
+		},
+	)
+	//!-vec
+	//!+splitcfg
+	splitter := textsplitter.NewRecursiveCharacter(
+		textsplitter.WithChunkSize(200),
+		textsplitter.WithChunkOverlap(20),
+	)
+	//!-splitcfg
+	//!+indexflow
+	genkit.DefineFlow(
+		"indexMenu",
+		func(ctx context.Context, path string) (any, error) {
+			// Extract plain text from the PDF.
+			pdfText, err := genkit.Run(ctx, "extract", func() (string, error) {
+				return readPdf(path)
+			})
+			if err != nil {
+				return nil, err
+			}
+
+			// Split the text into chunks.
+			docs, err := genkit.Run(ctx, "chunk", func() ([]*ai.Document, error) {
+				chunks, err := splitter.SplitText(pdfText)
+				if err != nil {
+					return nil, err
+				}
+
+				var docs []*ai.Document
+				for i := range len(chunks) {
+					docs = append(docs, ai.DocumentFromText(chunks[i], nil))
+				}
+				return docs, nil
+			})
+			if err != nil {
+				return nil, err
+			}
+
+			// Add chunks to the index.
+			err = menuPdfIndexer.Index(ctx, &ai.IndexerRequest{
+				Documents: docs, Options: nil,
+			})
+			return nil, err
+		},
+	)
+	//!-indexflow
+	
+	genkit.Init(ctx, nil)
+}
+
+//!+readpdf
+// Helper function to extract plain text from a PDF. Excerpted from 
+// https://github.com/ledongthuc/pdf
+func readPdf(path string) (string, error) {
+	f, r, err := pdf.Open(path)
+    defer f.Close()
+	if err != nil {
+		return "", err
+	}
+	
+	buf := bytes.Buffer{}
+    b, err := r.GetPlainText()
+    if err != nil {
+        return "", err
+    }
+    buf.ReadFrom(b)
```

Comment:
Not sure what's going on here. if `b` is a []byte, then you can just do `string(b)`.